### ISSUE TITLE
Strategy for operator precedence

### ISSUE THREAD

**Issue by Kuniwak** [Kuniwak/jsdoctypeparser#34]
### Problem

In v2.0.0-alpha-5, jsdoctypeparser do not have something to control operator precedence.

Related: #32 

### Approach

1. When given the type expression:

    ```
    op1 op2 OPERAND op3 op4
    ```

2. We will get the 2 groups of tokens:

    ```javascript
    var prefixOps = [op2, op1];
    var suffixOps = [op3, op4];
    ```

3. Mix tokens groups (in the example, prefix token have higher priority):

    ```javascript
    var mixedOps = flatten(zip(prefixOps, suffixOps))
    // [op2, op3, op1, op4]
    ```

4. Given a strategy function and sort by using it:

    ```javascript
    mixedOps.sort(strategyFn);
    // [op3, op2, op1, op4]
    ```

Maybe it is useful that operator tokens have own priority level and `strategyFn` care only the level.

Related codes: https://github.com/Kuniwak/jsdoctypeparser/blob/793483b1556974ead42e1514d762001680211d6f/peg_src/jsdoctype.pegjs#L52

**Comment by Kuniwak** [comment]
No, this approach is not perfect.
This approach cannot change priority for `UNION` or `GENERIC`.

**Comment by Kuniwak** [comment]
The  class of `UNION` and `GENERIC` is not equivalent to the postfix modifier's one.
Because the last one keep own context, but not the first one.
So, these classes should be splitted.

I think that this approach can resolve the priority problem about `UNION` and `GENERIC`.

**Comment by Kuniwak** [comment]
This is a PoC, but I feel that it is a nasty way since I read [the article](https://github.com/PhilippeSigaud/Pegged/wiki/PEG-Basics#an-example-grammar)...

```
{
  var OpPrecedence = {
    p1: 0,
    s1: 1,
    p2: 2,
    s2: 3,

    null: Number.MAX_VALUE,
  };

  function getOpPrecedence(op) {
    return OpPrecedence[String(op)];
  }

  function sortUnaryOps(prefixOps, operand, suffixOps, tokens) {
    var result = operand;
    var prefixOp = prefixOps.length > 0 ? prefixOps.pop() : null;
    var suffixOp = suffixOps.length > 0 ? suffixOps.shift() : null;

    while (prefixOp || suffixOp) {
      if (getOpPrecedence(prefixOp) < getOpPrecedence(suffixOp)) {
        result = [prefixOp, result];

        prefixOp = prefixOps.length > 0
          ? prefixOps.pop()
          : null;
      }
      else {
        result = [suffixOp, result];

        suffixOp = suffixOps.length > 0
          ? suffixOps.pop()
          : null;
      }
    }
    return result;
  }
}

expr = preOps:prefixUnaryOperator* operand:operand sufOps:suffixUnaryOperator* {
  return sortUnaryOps(preOps, operand, sufOps);
}

operand = "o"

prefixUnaryOperator = prefixUnaryOperator1
                    / prefixUnaryOperator2
suffixUnaryOperator = suffixUnaryOperator1
                    / suffixUnaryOperator2

prefixUnaryOperator1 = "p1"
prefixUnaryOperator2 = "p2"
suffixUnaryOperator1 = "s1"
suffixUnaryOperator2 = "s2"
```

**Comment by Kuniwak** [comment]
I felt that this approach is better..

```
expr = expr4

expr1 = p:prefixUnaryOperator1? x:operand {
  return p ? [p, x] : p;
}

expr2 = x:expr1 s:suffixUnaryOperator1? {
  return s ? [s, x] : x;
}

expr3 = p:prefixUnaryOperator2? x:expr2 {
  return p ? [p, x] : x;
}

expr4 = x:expr3 s:suffixUnaryOperator2? {
  return s ? [s, x] : x;
}

operand = "o"

prefixUnaryOperator1 = "p1"
prefixUnaryOperator2 = "p2"
suffixUnaryOperator1 = "s1"
suffixUnaryOperator2 = "s2"
```

**Comment by Kuniwak** [comment]
I started to implement the last approach.

Now, I learned the naive implementation (#37) cannot parse complex syntax that can cause left-recursion (for example, `Some[]![]`).

**Comment by Kuniwak** [comment]
[Expected operator precedence matrix](https://docs.google.com/spreadsheets/d/1_zR99ko_lgOGuNfvKsp6uzNCBWpDPSCWAzpWHtsGx9E/edit?usp=sharing).

**Comment by Kuniwak** [comment]
This problem was fixed on #37
